[lldb][progress] Improve Swift progress reporting

chelcassanova opened the pull request:
LLDB's progress reporting infrastructure has had the ability to add details to an overall progress report (added in
https://reviews.llvm.org/D143690) but this had yet to be implemented in existing progress reports. This commit adds this functionality to progress reports made for Swift operations (such as caching user imports) so that instead of sending several individual progress reports for each step in these operations, one progress report is created and then incrementally updated.

Importing Swift modules takes place in the Swift compiler which uses a callback function so that LLDB can perform its progress report. This originally called into a function in `lldb_private::SwiftASTContext` but will instead use a lambda function to increment a locally created progress report.

Related to https://github.com/apple/swift/pull/69730

rdar://105286354

chelcassanova committed:
[lldb][progress] Improve Swift progress reporting

LLDB's progress reporting infrastructure has had the ability to add
details to an overall progress report (added in
https://reviews.llvm.org/D143690) but this had yet to be implemented in
existing progress reports. This commit adds this functionality to
progress reports made for Swift operations (such as caching user
imports) so that instead of sending several individual progress reports
for each step in these operations, one progress report is created and
then incrementally updated.

Importing Swift modules takes place in the Swift compiler which uses a
callback function so that LLDB can perform its progress report. This
originally called into a function in `lldb_private::SwiftASTContext` but
will instead use a lambda function to increment a locally created
progress report.

rdar://105286354

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3263,7 +3281,20 @@ swift::ModuleDecl *SwiftASTContext::GetModule(const SourceModule &module,
   auto import_diags = getScopedDiagnosticConsumer();
 
   // Perform the import.
```
nit: Technically the import happens in the `ast->getModuleByName` call, so I'd move the progress setup above the comment.

chelcassanova replied:
Good point, I'll change this.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3881,7 +3908,7 @@ void SwiftASTContext::ValidateSectionModules(
     // We have to increment the completion value even if we can't get the module
     // object to stay in-sync with the total progress reporting.
-    progress.Increment(++completion);
+    progress.Increment(++completion, llvm::formatv("{0}", module.GetFileSpec().GetFilename().AsCString()));
```
No need for `llvm::formatv` here, you can just pass the call to `AsCString()`.

medismailben replied:
I guess if you remove this, you can also drop the `#include "llvm/Support/FormatVariadic.h"`

chelcassanova replied:
Oh nice, I was using the formatters from how the progress reports were initially created so I'll change this.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -8339,7 +8363,7 @@ bool SwiftASTContextForExpressions::CacheUserImports(
   
   for (const auto &attributed_import : src_file_imports) {
-    progress.Increment(++completion);
+    progress.Increment(++completion, llvm::formatv("{0}", source_file.getFilename().str()));
```
ditto.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -8452,14 +8476,10 @@ bool SwiftASTContext::GetCompileUnitImportsImpl(
   size_t completion = 0;
   for (const SourceModule &module : cu_imports) {
-    progress.Increment(++completion);
+    progress.Increment(++completion, llvm::formatv("{0}", module.path.back()));
```
ditto.

medismailben commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/TestSwiftProgressReporting.py:
```diff
@@ -8,26 +8,28 @@
         self.broadcaster = self.dbg.GetBroadcaster()
-        self.listener = lldbutil.start_listening_from(self.broadcaster,
-                                        lldb.SBDebugger.eBroadcastBitProgress)
+        self.listener = lldbutil.start_listening_from(
```
Is this your own reformatting or `black` output ? If it's the former, we should avoid moving code around for nothing as it rewrites history.

chelcassanova replied:
This was from `black`'s output after writing to the file, I can change it back to its original formatting if needed.

medismailben commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/TestSwiftProgressReporting.py:
```diff
@@ -8,26 +8,28 @@
 
-class TestSwiftProgressReporting(TestBase):
 
+class TestSwiftProgressReporting(TestBase):
```
What changed here ?

chelcassanova replied:
Nothing, I didn't make any changes here. Maybe a git issue from when I created the commit?

medismailben replied:
I'd revert these changes and see if the formatting test job fails on the PR.

medismailben commented on lldb/test/API/functionalities/progress_reporting/swift_progress_reporting/TestSwiftProgressReporting.py:
```diff
@@ -8,26 +8,28 @@
 
-        target, process, thread, _ = lldbutil.run_to_source_breakpoint(self, 'break here',
-                                          lldb.SBFileSpec('main.swift'))
+        target, process, thread, _ = lldbutil.run_to_source_breakpoint(
```
Same here.

chelcassanova replied:
I think this is also a result of `black` formatting.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1989,12 +1991,23 @@ SwiftASTContext::CreateInstance(lldb::LanguageType language, Module &module,
+    swift_ast_sp->m_ast_context_ap->SetPreModuleImportCallback(
+      [&progress](llvm::StringRef module_name, bool is_overlay) {
+        if (progress) {
+          progress->Increment(1, module_name.str());
```
Seems like we're not using the `is_overlay` argument anymore. @adrian-prantl do you think this is relevant information to the user or should we just get rid of it in the callback ?

adrian-prantl replied:
What do you mean be "_we_ are not"? There is no overlay for the Swift stdlib module since it's implemented in Swift.

medismailben replied:
This was not the best example: replied here were it makes more sense: https://github.com/apple/llvm-project/pull/7769#discussion_r1387283810

medismailben reviewed (commented):
Nice! This just need some polishing and it should be good to go 🙂

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1989,12 +1991,23 @@ SwiftASTContext::CreateInstance(lldb::LanguageType language, Module &module,
   {
     LLDB_SCOPED_TIMERF("%s (getStdlibModule)", m_description.c_str());
     const bool can_create = true;
+    std::unique_ptr<lldb_private::Progress> progress = std::make_unique<lldb_private::Progress>("Importing Swift standard library modules");
```
`module`. AFAIK there is only one `Swift` module that is being imported there.

medismailben replied:
But IIUC, the swift standard library module can have dependencies that are also loaded implicitly right ?

chelcassanova replied:
This works when the progress reports have details (e.g. `Importing Swift module: Foundation`) but I think it would make the overall title a little less clear (i.e. having one main progress report that says `Importing Swift module` even though multiple individual imports are about to be performed). I can change it but maybe we also have an overall title and then an `Increment` title in addition to having `details`?

medismailben replied:
I agree with Chelsea here: From a user perspective, it will look like we're only importing a single module, where actually that one module import trigger many other modules to be imported. I think `modules` makes more sense here.

adrian-prantl replied:
Swift has 1 dependency: SwiftShims, which isa Clang module. So I guess, you're right.

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -2479,8 +2492,21 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
+      [&progress](llvm::StringRef module_name, bool is_overlay) {
+        if (progress) {
+          progress->Increment(1, module_name.str());
+          return true;
```
Out of curiosity: what does the return value do?
// shorter:
if (progress)
   progress->Increment(1, module_name.str());
return progress;

chelcassanova replied:
I think I had it this way because I thought the callback had to return a bool but I'm mistaken, I'll change it to your shorter way.

adrian-prantl replied:
I would suggest changing the callback to `std::function<void(...)>` if the return value isn't used there.

chelcassanova replied:
Noted, will change this

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -8318,10 +8345,7 @@ bool SwiftASTContextForExpressions::CacheUserImports(
-                                  source_file.getFilename().data()),
-                    src_file_imports.size());
-
+  Progress progress("Caching Swift user imports from");
```
This is very LLDB-centric language:
"Importing module used in expression"

chelcassanova replied:
I can change this, but do we still want to tell users that Swift imports were cached?

adrian-prantl replied:
No, the existing wording confused even me (as someone familiar with the implementation) and I had to look at the code to understand what "caching" in this context is even supposed to mean.

chelcassanova replied:
Interesting, for all progress reports I used the titles from the original implementation. If we're also telling the user that we're importing a module due to the use of an expression do we want to tell them what the expression that triggered the import was in the report? (this might be a bit overboard, I'm not sure if we even have the expression itself at this level)

medismailben replied:
I don't think we can achieve that today since progress report can't be related one another. You can stack them but that don't necessarily means the one at the bottom of the stack is a parent of the one above. It would be a nice improvement for later though :)

chelcassanova replied:
Yeah it would be nice to have this in the future 👍🏾

chelcassanova commented:
https://github.com/apple/swift/pull/69730
@swift-ci test

adrian-prantl commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -8452,14 +8476,10 @@ bool SwiftASTContext::GetCompileUnitImportsImpl(
-                    compile_unit->GetPrimaryFile().GetFilename()),
-      cu_imports.size());
-
+  Progress progress("Getting Swift compile unit imports");
```
Suggestion: "Importing modules used by compilation unit {0}", compile_unit->GetPrimaryFile().GetFilename()

chelcassanova replied:
Good suggestion, I'll change this.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3263,7 +3281,20 @@ swift::ModuleDecl *SwiftASTContext::GetModule(const SourceModule &module,
   // Perform the import.
+  std::unique_ptr<lldb_private::Progress> progress = std::make_unique<lldb_private::Progress>("Importing Swift modules");
+  ast->SetPreModuleImportCallback(
+    [&progress](llvm::StringRef module_name, bool is_overlay) {
```
@adrian-prantl My bad, the swift standard library module was not a good example. Here however, we could load modules are overlay right ? Do you think that the overlay part is relevant information for the user and print the following for instance:
```
Importing Swift Module: Darwin (overlay).
```
Or do you think it's not necessary to mention that it's an overlay:
```
Importing Swift Module: Darwin.
```

adrian-prantl replied:
Yes, because as your example illustrates, they come in pairs. A Clang module, and it's similarly named Swift overlay.

medismailben replied:
Thank you for clarifying :)

chelcassanova committed:
Update progress titles, add strings for importing overlay modules

Changes the titles of some progress reports to better reflect what
operation is being done, updates the details sections when
importing Swift modules to also report when Swift overlay modules are
being imported and removes the boolean return types from the lambda
functions used to increment progress reports.

Also updates the API test to revert the formatting done by black and
updates the progress titles checked by the test.

chelcassanova committed:
fixup! Update progress titles, add strings for importing overlay modules

Remove unused import

chelcassanova commented:
https://github.com/apple/swift/pull/69730
@swift-ci smoke test

chelcassanova commented:
https://github.com/apple/swift/pull/69730
@swift-ci please test

chelcassanova committed:
Use correct filename for expression module import progress report

Progress reports for importing a module used in an expression used the
wrong variable to obtain the module's filename which caused the progress
report to appear incorrect. This updates the filename used to be correct.

augusto2112 commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1989,12 +1990,21 @@ SwiftASTContext::CreateInstance(lldb::LanguageType language, Module &module,
   {
     LLDB_SCOPED_TIMERF("%s (getStdlibModule)", m_description.c_str());
     const bool can_create = true;
+    std::unique_ptr<lldb_private::Progress> progress = std::make_unique<lldb_private::Progress>("Importing Swift standard library modules");
```
Why do some of these progress options need to be wrapped in a `std::unique_ptr`?

chelcassanova replied:
Leftover from a previous implementation before the PreImport setter was added, this can be changed to just use the progress object itself instead of the unique pointer since it will be destroyed when it goes out of scope anyways.

chelcassanova committed:
Clear callback function on scope exit

Adds a call to `llvm::make_scope_exit` to clear the callback function by
setting it to an empty lambda function to prevent an out-of-scope access
to the progress variable.

chelcassanova commented:
https://github.com/apple/swift/pull/69730
@swift-ci please test

chelcassanova committed:
Remove use of unique pointer for module loading progress objects

The progress report objects used for Swift module importing used unique
pointers as a leftover from a previous implementation. This commit
removes them and changes them to regular `Progress` objects. Also
corrects how overlay imports were reported in their update string as the
logic to add the "(overlay)" word was reversed in their ternary statements.

JDevlieghere commented on lldb/source/Core/Progress.cpp:
```diff
@@ -30,10 +30,9 @@ Progress::~Progress() {
     m_completed = m_total;
-    ReportProgress();
-  }
+  ReportProgress();
```
Shouldn't this go upstream?

medismailben replied:
@chelcassanova This should go upstream, before you land this patch otherwise, you could get a merge conflict.

medismailben replied:
This depends on llvm/llvm-project#73605. @chelcassanova make sure to cherrypick here the merge commit after you land it upstream ;)

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1989,6 +1990,20 @@ SwiftASTContext::CreateInstance(lldb::LanguageType language, Module &module,
     const bool can_create = true;
+
+    // Report progress on module importing by using a callback function in swift::ASTContext
+    Progress progress("Importing Swift standard library modules");
```
Are there multiple standard library modules or is there one? I know it might be pulling in other modules as dependencies. I guess we could sidestep this by omitting "modules" at the end. @adrian-prantl WDYT?

medismailben replied:
ping @chelcassanova

chelcassanova replied:
"Modules" can be omitted here for clarity

medismailben replied:
I don't understand ... you say it can be omitted but it's still in the PR 🧐 ...

```suggestion
    Progress progress("Importing Swift standard library");
```

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1989,6 +1990,20 @@ SwiftASTContext::CreateInstance(lldb::LanguageType language, Module &module,
+
+    // Clear the callback function on scope exit to prevent an out-of-scope access of the progress local variable
+    auto on_exit = llvm::make_scope_exit([&](){
+      swift_ast_sp->m_ast_context_ap->SetPreModuleImportCallback([](llvm::StringRef module_name, bool is_overlay) {});
```
Does `swift_ast_sp->m_ast_context_ap->SetPreModuleImportCallback({})` work?

chelcassanova replied:
I tried it locally and it does, but per [Augusto's comment](https://github.com/apple/llvm-project/pull/7769#discussion_r1396471773) we might want to set the callback function to a generic progress report in the event that a module import is triggered outside of these functions.

JDevlieghere commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -2479,6 +2494,20 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
+      [&progress](llvm::StringRef module_name, bool is_overlay) {
+        progress.Increment(1, (is_overlay ? module_name.str() + " (overlay)" : module_name.str()));
+      }
+);
```
Needs to be clang-formatted.

augusto2112 commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1989,6 +1990,20 @@ SwiftASTContext::CreateInstance(lldb::LanguageType language, Module &module,
+
+    // Clear the callback function on scope exit to prevent an out-of-scope access of the progress local variable
+    auto on_exit = llvm::make_scope_exit([&](){
+      swift_ast_sp->m_ast_context_ap->SetPreModuleImportCallback([](llvm::StringRef module_name, bool is_overlay) {});
```
If I understand correctly, today we'll always emit a generic progress of "Importing [overlay] module X" whenever a module is imported. With this change, we'll have better progress reporting messages, but won't report anything else that happens to be imported outside of these functions. Perhaps re-introduce the generic progress message in case modules are imported outside these functions?

chelcassanova replied:
Good point, I hadn't thought about the scenario of module imports being triggered in LLDB outside of these functions. If that does happen and the callback function is generic then LLDB might not be able to give a detailed progress report (since the generic callback wouldn't have the context of which modules were imported or where the import took place from) but it still indicates that LLDB caused a Swift module import which the user would want to know.

chelcassanova committed:
Apply proper formatting

`clang-format`s previous changes

chelcassanova committed:
Add generic progress report callback

Reuses `ReportModuleImportCallback` as a generic callback function that gets set upon scope exit instead
of using a blank function so that Swift module imports are reported when
the import is triggered outside of `SwiftASTContext::GetModule` and `SwiftASTContext::CreateInstance`.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3205,10 +3240,7 @@ void SwiftASTContext::CacheModule(swift::ModuleDecl *module) {
                                                   bool is_overlay) {
-  Progress progress(llvm::formatv(is_overlay ? "Importing overlay module {0}"
-                                             : "Importing module {0}",
-                                  module_name)
-                        .str());
+  Progress progress("Importing Swift module");
   return true;
 }
```
AFAIK, this is a no-op now. You can get rid of it.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3262,6 +3294,20 @@ swift::ModuleDecl *SwiftASTContext::GetModule(const SourceModule &module,
+  // Clear the callback function on scope exit to prevent an out-of-scope access
+  // of the progress local variable
+  auto on_exit = llvm::make_scope_exit(
+      [&]() { ast->SetPreModuleImportCallback(ReportModuleLoadingProgress); });
```
Can you try setting the callback to an empty lambda instead of reusing `ReportModuleLoadingProgress` ?

chelcassanova replied:
I'm using a generic callback here to report on any module imports progress reports not caught by the current callback lambdas but this can be replaced with a generic module import progress report.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -8452,14 +8494,11 @@ bool SwiftASTContext::GetCompileUnitImportsImpl(
   if (cu_imports.size() == 0)
     return true;
 
-  Progress progress(
-      llvm::formatv("Getting Swift compile unit imports for '{0}'",
-                    compile_unit->GetPrimaryFile().GetFilename()),
-      cu_imports.size());
-
+  Progress progress("Importing module used by compile unit");
   size_t completion = 0;
   for (const SourceModule &module : cu_imports) {
-    progress.Increment(++completion);
+    progress.Increment(
+        ++completion, compile_unit->GetPrimaryFile().GetFilename().AsCString());
```
I think this is wrong: You just moved the use of `compile_unit` in the loop to use it in the message detail but it never changes. I think you should leave the progress message the same it was and use the `SourceModule &module` variable to give extra information in the detailed messaged.

chelcassanova replied:
Ah, I didn't notice that `compile_unit` was the wrong variable to use here. @adrian-prantl made the suggestion to say "Importing module used by compile unit..." but maybe we could say something like "Importing module used by compile unit <compile unit name>: <module>" with the module as the detail but that might be overkill

medismailben replied:
Why would that be overkill ?

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -293,6 +293,9 @@ class SwiftASTContext : public TypeSystemSwift {
   swift::ModuleDecl *CreateModule(const SourceModule &module, Status &error,
                                   swift::ImplicitImportInfo importInfo);
 
+  /// Generic callback function used for progress reporting that gets
+  /// invoked by the Swift compiler and gets set upon scope exit when
+  /// a more detailed progress report was used as the callback
   static bool ReportModuleLoadingProgress(llvm::StringRef module_name,
                                           bool is_overlay);
```
I think we can just delete this if you replace it by an empty lambda.

medismailben reviewed (changes requested):
There are few things that need to change before this lands.

chelcassanova committed:
Add generic progress report lambda, update compile unit import progress message

Instead of clearing the pre module import callback with
`ReportModuleLoadingProgress`, clear it with a lambda that uses a
generic message to report Swift module importing that gets triggered
outside of the places where Swift import progress reports are currently
being made.

Also, the progress title for Swift compile unit imports has been fixed
and updated to show both the current compile unit and the module being
imported, and the test has been updated to reflect this.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -2479,6 +2499,25 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
+
+    // Report progress on module importing by using a callback function in
+    // swift::ASTContext
+    Progress progress("Importing Swift standard library modules");
```
Same here, drop the "modules" part.

```suggestion
    Progress progress("Importing Swift standard library");
```

chelcassanova replied:
Not sure how I missed both of these, thanks for the correction!

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -2479,6 +2499,25 @@ lldb::TypeSystemSP SwiftASTContext::CreateInstance(
+    auto on_exit = llvm::make_scope_exit([&]() {
+      swift_ast_sp->m_ast_context_ap->SetPreModuleImportCallback(
+          [](llvm::StringRef module_name, bool is_overlay) {
+            Progress progress("Importing Swift modules");
```
You can just call the constructor here, I don't think you need to create a `progress` variable.

```suggestion
            Progress("Importing Swift modules");
```

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3262,6 +3292,24 @@ swift::ModuleDecl *SwiftASTContext::GetModule(const SourceModule &module,
+  auto on_exit = llvm::make_scope_exit([&]() {
+    ast->SetPreModuleImportCallback(
+        [](llvm::StringRef module_name, bool is_overlay) {
+          Progress progress("Importing Swift modules");
```
ditto.

```suggestion
          Progress("Importing Swift modules");
```

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3881,7 +3925,8 @@ void SwiftASTContext::ValidateSectionModules(
     // We have to increment the completion value even if we can't get the module
     // object to stay in-sync with the total progress reporting.
-    progress.Increment(++completion);
+    progress.Increment(++completion,
+                       module.GetFileSpec().GetFilename().AsCString());
```
This looks wrong, you need to re-use the `module_name` from the loop range declaration.

```suggestion
    progress.Increment(++completion, module_name);
```

chelcassanova replied:
Ah, I reused the original implementation grabbed the module name from its FileSpec so that's how I missed this

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3868,11 +3916,7 @@ void SwiftASTContext::ValidateSectionModules(
-                    module.GetFileSpec().GetFilename().AsCString()),
-      module_names.size());
-
+  Progress progress("Loading Swift module", module_names.size());
```
We should experiment and see changing the progress title here is actually a good thing. Now you don't know what parent module are we loading the submodule for.

chelcassanova replied:
`Progress progress("Loading Swift module for parent module <parent_module_name>", module_names.size());` maybe?

chelcassanova replied:
Or `"Loading Swift module '<parent_module>'. Submodule: <details>`? The only thing is that the title would look empty when first displayed.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -8318,10 +8363,7 @@ bool SwiftASTContextForExpressions::CacheUserImports(
-                                  source_file.getFilename().data()),
-                    src_file_imports.size());
-
+  Progress progress("Importing module used in expression");
```
```suggestion
  Progress progress("Importing modules used in expression");
```

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.h:
```diff
@@ -293,6 +293,9 @@ class SwiftASTContext : public TypeSystemSwift {
   swift::ModuleDecl *CreateModule(const SourceModule &module, Status &error,
                                   swift::ImplicitImportInfo importInfo);
 
+  /// Generic callback function used for progress reporting that gets
+  /// invoked by the Swift compiler and gets set upon scope exit when
+  /// a more detailed progress report was used as the callback
   static bool ReportModuleLoadingProgress(llvm::StringRef module_name,
                                           bool is_overlay);
```
This is still here ...

chelcassanova replied:
Removed, I thought I took this out of the header 😓

medismailben reviewed (commented):
Getting close but there are still a few things to change.

chelcassanova committed:
[lldb][progress] Always report progress upon Progress object destruction (#73605)

This commit allows a final progress report upon the destruction of the
`Progress` object to happen at all times as opposed to when the progress
was not completed.

(cherry picked from commit c846f8ba104f18053ce666de7e2da0a82deb45cd)

chelcassanova committed:
Revert in order to cherry-pick from upstream

chelcassanova committed:
Clean up progress titles

Retitles some progress reports to be more accurate and updates the test
to match, removes the use of a
local variable in the generic lambda function that clears the callback,
remove `ReportModuleLoadingProgress` from SwiftASTContext's header file.

medismailben commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -3916,7 +3916,10 @@ void SwiftASTContext::ValidateSectionModules(
 
-  Progress progress("Loading Swift module", module_names.size());
+  Progress progress(
+      llvm::formatv("Loading Swift module '{0}'. Submodule",
```
Sorry for going back and forth on this but "swift module <modulename> submodule" sounds too repetitive. If we retitle it this way it addresses the repetitiveness.

```suggestion
      llvm::formatv("Loading Swift module '{0}' dependencies",
```

chelcassanova replied:
This is much better wording, I'll change the title here

medismailben reviewed (approved):
LGTM with a comment! Thanks for your patience 😄

chelcassanova committed:
Fix test and change progress title

Fixes API test so that it passes and changes the title of the progress
for loading Swift modules for clarity

JDevlieghere reviewed (approved):
LGTM

chelcassanova commented:
https://github.com/apple/swift/pull/69730
@swift-ci please test

chelcassanova commented:
https://github.com/apple/swift/pull/69730
@swift-ci test windows platform
